### ISSUE TITLE
sf causes a segfault or failed connections for odbc when loaded before running dbConnect in some environments

### ISSUE THREAD

**Issue by botanize** [r-spatial/sf#1742]
I'm not sure if this is a `knitr`, `odbc` or `sf` issue, but it happens whenever `sf` is imported (maybe just if `sf` is imported before executing `dbConnect`), so I'm going to start here.

Knitting this document fails:

    ---
    output: html_document
    ---
    
    ```{r}
    library(odbc)
    library(sf)
    
    con = dbConnect(odbc(), "GISLibrary")
    dbDisconnect(con)
    ```

Producing the following output:

```
processing file: knitr_db_test1.Rmd

 *** caught segfault ***
address 0x7f975f481cb0, cause 'memory not mapped'

Traceback:
 1: odbc_connect(connection_string, timezone = timezone, timezone_out = timezone_out,     encoding = encoding, bigint = bigint, timeout = timeout)
 2: OdbcConnection(dsn = dsn, ..., timezone = timezone, timezone_out = timezone_out,     encoding = encoding, bigint = bigint, timeout = timeout,     driver = driver, server = server, database = database, uid = uid,     pwd = pwd, dbms.name = dbms.name, .connection_string = .connection_string)
 3: .local(drv, ...)
 4: dbConnect(odbc(), "GISLibrary")
 5: dbConnect(odbc(), "GISLibrary")
 6: eval(expr, envir, enclos)
 7: eval(expr, envir, enclos)
 8: withVisible(eval(expr, envir, enclos))
 9: withCallingHandlers(withVisible(eval(expr, envir, enclos)), warning = wHandler,     error = eHandler, message = mHandler)
10: handle(ev <- withCallingHandlers(withVisible(eval(expr, envir,     enclos)), warning = wHandler, error = eHandler, message = mHandler))
11: timing_fn(handle(ev <- withCallingHandlers(withVisible(eval(expr,     envir, enclos)), warning = wHandler, error = eHandler, message = mHandler)))
12: evaluate_call(expr, parsed$src[[i]], envir = envir, enclos = enclos,     debug = debug, last = i == length(out), use_try = stop_on_error !=         2L, keep_warning = keep_warning, keep_message = keep_message,     output_handler = output_handler, include_timing = include_timing)
13: evaluate::evaluate(...)
14: evaluate(code, envir = env, new_device = FALSE, keep_warning = !isFALSE(options$warning),     keep_message = !isFALSE(options$message), stop_on_error = if (is.numeric(options$error)) options$error else {        if (options$error && options$include)             0L        else 2L    }, output_handler = knit_handlers(options$render, options))
15: in_dir(input_dir(), evaluate(code, envir = env, new_device = FALSE,     keep_warning = !isFALSE(options$warning), keep_message = !isFALSE(options$message),     stop_on_error = if (is.numeric(options$error)) options$error else {        if (options$error && options$include)             0L        else 2L    }, output_handler = knit_handlers(options$render, options)))
16: eng_r(options)
17: block_exec(params)
18: call_block(x)
19: process_group.block(group)
20: process_group(group)
21: withCallingHandlers(if (tangle) process_tangle(group) else process_group(group),     error = function(e) {        setwd(wd)        cat(res, sep = "\n", file = output %n% "")        message("Quitting from lines ", paste(current_lines(i),             collapse = "-"), " (", knit_concord$get("infile"),             ") ")    })
22: process_file(text, output)
23: knitr::knit(knit_input, knit_output, envir = envir, quiet = quiet)
24: rmarkdown::render("/Users/reidjp/Documents/Projects/scratchpad/knitr_db_test1.Rmd",     encoding = "UTF-8")
An irrecoverable exception occurred. R is aborting now ...
```

Comment out `library(sf)` and it works. Replace `library(sf)` with `library(tigris)` and it fails.

To make matters more interesting, this works:

    ---
    title: "DB Test"
    ---

    ```{r}
    library(odbc)
    con = dbConnect(odbc(), 'GISLibrary')
    dbGetQuery(con, "SELECT 1 as num")
    dbDisconnect(con)
    ```

    ## Load sf, retry query
    ```{r}
    library(sf)
    con = dbConnect(odbc(), 'GISLibrary')
    dbGetQuery(con, "SELECT 1 as num")
    dbDisconnect(con)
    ```

**Comment by edzer** [comment]
I have no clue. @etiennebr ?

**Comment by botanize** [comment]
It looks like I can remove knitr from the equation. I get the same behavior with the command line tools `R` or `Rscript`:

This works:

    Rscript -e "library(odbc); con = dbConnect(odbc(), 'GISLibrary'); dbGetQuery(con, 'select 1 num'); dbDisconnect(con);"

This doesn't:

    Rscript -e "library(odbc); library(sf);  con = dbConnect(odbc(), 'GISLibrary'); dbGetQuery(con, 'select 1 num'); dbDisconnect(con);"

**Comment by edzer** [comment]
What is your `sessionInfo()` after loading both? Have you tried reinstalling `sf`?

**Comment by edzer** [comment]
Have you looked at #1743?

**Comment by botanize** [comment]
Same behavior after upgrading from sf 1.0 to 1.0-2.

Session Info after just `odbc`:

    R version 4.1.0 (2021-05-18)
    Platform: x86_64-apple-darwin17.0 (64-bit)
    Running under: macOS Big Sur 10.16

    Matrix products: default
    BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
    LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

    locale:
    [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

    attached base packages:
    [1] stats     graphics  grDevices utils     datasets  methods   base     

    other attached packages:
    [1] odbc_1.3.2

    loaded via a namespace (and not attached):
     [1] bit_4.0.4       compiler_4.1.0  ellipsis_0.3.2  hms_1.1.0      
     [5] DBI_1.1.1       Rcpp_1.0.7      bit64_4.0.5     vctrs_0.3.8    
     [9] blob_1.2.2      lifecycle_1.0.0 pkgconfig_2.0.3 rlang_0.4.11

Session info after loading `odbc` then `sf`:

    Linking to GEOS 3.8.1, GDAL 3.2.1, PROJ 7.2.1
    R version 4.1.0 (2021-05-18)
    Platform: x86_64-apple-darwin17.0 (64-bit)
    Running under: macOS Big Sur 10.16

    Matrix products: default
    BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
    LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

    locale:
    [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

    attached base packages:
    [1] stats     graphics  grDevices utils     datasets  methods   base     

    other attached packages:
    [1] sf_1.0-2   odbc_1.3.2

    loaded via a namespace (and not attached):
     [1] Rcpp_1.0.7         magrittr_2.0.1     hms_1.1.0          units_0.7-2       
     [5] tidyselect_1.1.1   bit_4.0.4          R6_2.5.0           rlang_0.4.11      
     [9] fansi_0.5.0        blob_1.2.2         dplyr_1.0.7        tools_4.1.0       
    [13] grid_4.1.0         KernSmooth_2.23-20 utf8_1.2.2         e1071_1.7-7       
    [17] DBI_1.1.1          ellipsis_0.3.2     class_7.3-19       assertthat_0.2.1  
    [21] bit64_4.0.5        tibble_3.1.3       lifecycle_1.0.0    crayon_1.4.1      
    [25] purrr_0.3.4        vctrs_0.3.8        glue_1.4.2         proxy_0.4-26      
    [29] compiler_4.1.0     pillar_1.6.2       generics_0.1.0     classInt_0.4-3    
    [33] pkgconfig_2.0.3

Just for "fun" I tested every package loaded via namespace, one-by-one, with odbc, none caused the connection to fail.

**Comment by botanize** [comment]
The issue persists after I uninstalled `Rcpp`, `sf` and `odbc`, restarted the session, then re-installed each.

**Comment by etiennebr** [comment]
A few thoughts and questions to try to help you since this is likely unique to your installation (I could not reproduce with my linux setup):
- What is the definition of the `GISLibrary` data source? Which Database is it? I would try to connect using explicit string. It is driver specific, but here's what it look like for [PostgreSQL](https://db.rstudio.com/databases/postgresql/):
``` r
con <- DBI::dbConnect(odbc::odbc(),
                      Driver   = "PostgreSQL Unicode",
                      Server   = "127.2.56.12",
                      Database = "gis",
                      Port     = 5432)
```
The [RStudio documentation](https://db.rstudio.com/databases/) is very complete if you need help to configure for your database. Try to connect to another database if you can; if I understand correctly, the segfault happens when `odbc`, then `sf`, then connection. 
- Can you skip `odbc` and use another driver? https://dbi.r-dbi.org/
- What happens if you load `sf` later?
```
Rscript -e "library(odbc); con = dbConnect(odbc(), 'GISLibrary');  library(sf);  dbGetQuery(con, 'select 1 num'); dbDisconnect(con);"
```

**Comment by botanize** [comment]
`GISLibrary` is a SQL Server database. We use DSNs and kerberos for single-sign-on authentication when connecting to databases. I have the exact same problem when I use UID/PWD authentication. I'd be surprised if the database connection parameters had anything to do with it, since it works fine as long as `sf` isn't loaded before the first connection attempt.

I've tried a lot of stuff and I can only get this to reproduce when I use the `odbc` package with the `FreeTDS` driver. Everything works as expected when I use the `odbc` package and the Microsoft `ODBC Driver 17 for SQL Server` or `PostgreSQL Unicode` driver, or any other DBI package.

I think I mentioned above that I can load `sf` after the first `dbConnect` call and everything works as expected, the problem only occurs with the `odbc` package, the `FreeTDS` odbc driver and when loading `sf` before the first `dbConnect` call.

I tried reverting to R 4.0, sf 0.9-5, odbc 1.2.3 and the problem persists.

    R(72974,0x11017ee00) malloc: *** error for object 0x7fded3416c00: pointer being freed was not allocated
    R(72974,0x11017ee00) malloc: *** set a breakpoint in malloc_error_break to debug

In RStudio, /usr/local/bin/R and /usr/local/bin/Rscript this only happens with `sf` loaded before the first `dbConnect` call. But, I also can't connect in `R.app`, even without loading `sf`, so I'll close this for now and check in with the R macOS developers to see if we can find the source of the problem.
